### ISSUE TITLE
Using Components<T> is not safe for structural mutation

### ISSUE THREAD

**Issue by Helco** [Doraku/DefaultEcs.Analyzer#10]
Using `Components<T>`  and also creating/modifying entities can lead to surprising bugs. Reallocations can destroy the correct link to the pool arrays and thus modifications are no longer applied.
This probably is also just a case of documentation, but the general feel for this is: even in single-threaded applications, even when always using buffers in all systems, using the recorder is the only safe way to apply structural modifications.

Minimal reproduction (expected to print `true`): 
```csharp
partial class SomeSystem : AEntitySetSystem<float>
{
    public SomeSystem(World world) : base(world, CreateEntityContainer, useBuffer: true)
    { }

    [Update]
    private void Update(float state, in Entity entity, ref bool byRef)
    {
        for (int i = 0; i < 10; i++)
            World.CreateEntity().Set(false);
        byRef = true;
    }

    public static void Main(string[] args)
    {
        var world = new World();
        var entityA = world.CreateEntity();
        entityA.Set(false);

        var system = new SomeSystem(world);
        system.Update(0f);

        Console.WriteLine(entityA.Get<bool>());
    }
}
```

**Comment by Doraku** [comment]
Yep `Components<T>` is only safe when doing no structural changes to entity as it caches the mapping and components arrays to remove one indirection compared to using `Entity.Get<T>()`. When you add new components `T` those arrays may need to be resized so the ones cached in `Components<T>` are no longer valid. I would expect the analyzer to produce a warning for the `Entity.Set<T>` operation in the update method in your example but I should also add more documentation on the type directly.

**Comment by Doraku** [comment]
Need some new analyzer to produce a warning for entity creation even if `userBuffer` is true.
